ops/model.py

```diff
@@ -176,6 +176,18 @@ def get_binding(self, binding_key: typing.Union[str, 'Relation']) -> 'Binding':
         """
         return self._bindings.get(binding_key)
 
+    def get_planned_unit_count(self) -> int:
```

Comment:
I kinda feel like this should be under the `Application` class, not the `Model` class? Doing `self.model.get_planned_unit_count` seems counter-intuitive, as it doesn't point to a specific application?

Also, I think I'd prefer `get_pending_units` - a little terser :)

Reply:
As a follow-up to this, I thought the current `goal-state` tool could indeed provide a list of "pending" units, rather than the total target number of units? 

Does this implementation provide a way to distinguish if indeed the other units are "pending" or already provisioned?

Reply:
Agree that this could live in Application rather than model. I'll make that change. 

I renamed to "planned" because "pending" isn't accurate. The units might be in a pending state, or they might be up and running. I added the _count at the end because we probably want to implement a method that gets planned units, including all the details, once we are more comfortable with goal state. This method would then just turn into a convenience wrapper that gets the count, without breaking existing code.

Reply:
... and to answer your second question. Yes! We can determine whether units are pending or deployed. I think that we wanted to avoid giving folks that information, though, keeping them focused just on "this application is going to have x units". Unless I misunderstood, in which case, I can refactor!

Reply:
Okay, so this method may well continue to be useful, but I think we need pending too, otherwise how will charmers make a decision about when to take action with a given unit? They'll need to do something like:

```python
def _on_peer_relation_joined(self, event):
     if self.app.get_pending_units() == 0:
        # do stuff
```

Saying that, I do wonder if for the basic use-case, knowing that there is more than 1 planned would allow a charmer to make decisions about commands to execute to enable HA or not, etc, etc, so perhaps we could add pending in a later PR. The main thing we wanted to avoid giving people was "There are two more units coming, and these are their IDs" - rather just "There are 2 more units coming (they may or may not ever materialise", which I guess this current PR does.

Reply:
I think this will get us there.

@dstathis was doing something similar in his Cassandra charm, and @jlounder has spoken about this before too.

Reply:
After sleeping on this, I have some thoughts. (I'm adding this to a spec in the Operator framework specs folder, for posterity.)

First off, here's an example of goal state, yoinked from the Juju core goal state tests:

```
units:
  mysql/0:
    status: active
    since: 2200-11-05 15:29:12Z
relations:
  db:
    mysql/0:
      status: active
      since: 2200-11-05 15:29:12Z
  server:
    wordpress/0:
      status: active
      since: 2200-11-05 15:29:12Z
```
Basically, goal state is just a list of units in this application, and relations to this application. It doesn't really belong in the Application or Unit classes, because you can't call goal state on another application -- you can only see the part of the model related directly to the charm executing the code. 

Since goal state is a special thing that has a specific relation to the current charm, it should probably either as a standalone class.

Here's a sketch of an implementation:

```
ops/model.py

class GoalState:
    def __init__(self, meta, backend):
        self._units = []
        self._relations = []
        self._backend = backend
        self.meta = meta

    def _fetch(self):
        state = self._backend.goal_state()  # TODO: cache?
        self._units = state.get('units', [])
        self._relations = state.get('relations', [])

    def peer_count(self):  # Tempted to @property, but I also want it to feel dynamic
        self._fetch()
        return len(self._units)

class ModelBackend:
    ...
    def goal_state(self):
        return self._run('goal-state', return_output=True, use_json=True)
```
We could then add the GoalState to the model, like we add the Application and Unit classes, and call it from CharmBase with `self.model.goal.peer_count()`

Reply:
Hmm. Interesting for a couple of reasons:

- Each unit has a status, meaning we can easily derive `planned_units` and `pending_units`, where planned units is `len(self._units)` and pending units is something line `len([u for u in self._units if u["status"] != "active")`, right?

- I'm still unconvinced this is a model level construct, and not an application level construct, precisely because you can't interrogate the goal state of *another application*, only yourself if that makes sense?

Reply:
Coming at it from another angle, I saw the "goal" of goal-state as "I'm going to be HA and I should be ready for that".

For the Juju core goal state of "a mysql, a webserver", the _overall_ state of the model probably makes sense. From an OF perspective, there are a couple of cases. They're all different from an application design perspective, and we cannot plan for all of them, but let's say we have the following cases:

* A Grafana 'cluster' which will want to know its end goal state so it does not bother initializing data in a local sqlite database, then immediately needing to shut down and migrate everything over to MySQL. Grafana needs to wait for a relation to MySQL in any case, which can be handled from the charm code itself, but instead of checking "do I have a relation to a mysql || do I have peer relations" and branching the logic out, it can interrogate "do I have more than one unit planned? If so, just wait on DB initialization until I have a MySQL"
* A Ceph cluster which "needs" to know how many initial nodes it should have. Sure, Ceph _can_ scale up and down, but it requires a bunch of twiddling to add/remove OSDs on the fly, and waiting for monitor initialization until all of the units are present is much easier.
* A Cassandra cluster which "needs" to know how many nodes will be present before initialization so data can be appropriate sharded. Same basic case as Ceph.
* A `etcd` cluster or anything which uses Raft which won't reach quorum without a minimum number of nodes
* A RabbitMQ cluster where node start ordering matters if the entire cluster goes down (this really doesn't apply to k8s charms, but still)
* Graylog "wants" both ElasticSearch and MongoDB, both of which "want" to have at least 3 nodes before they are "ready"

All of these cases are slightly different, but all align on a couple of points:

* Communication about when the overall application is "ready" to communicate to other charms to reach the **Juju** goal state rather than the **charm** goal state can be done by setting relation data when it is "ready"
* For charms which support iterative startup (like Rabbit through `rabbitmqctl`), this can be performed over peer relations, and, again, readiness communicated to external relations via `relation-set`

Yes, goal state can help avoid the "event storm" Leon referenced, but primarily, from my POV, by having whatever charm depends on its relation counterpart (be that Graylog waiting for an Elastic and Mongo, Grafana waiting for a MySQL) put itself into `WaitingStatus`. I used these two as examples because Graylog requires multiple external relations to achieve their goal state before it can be initialized, and Grafana needs to wait for a single one (assuming a single MySQL). 

From this perspective, we can avoid an "event storm" by, simply, not dumping a bunch of relation data over to external relations until quorum/initialization is complete. That's really up to charm authors, since they know when application X or Y is "ready" to talk to the outside world, and charms which require quorum before they can operate shouldn't be blindly firing off relation data on startup/relation-joined without checking whether they're operational first.

From an OF perspective, it's not reasonable to provide a complete abstraction for everything the entire application **model** may need, but we _can_ provide an abstraction for what a **single** application may need.

Whew, that got long. But in the end, I agree that this should probably be on `Application` for the reasons above.